flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/RecordWriterDelegate.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.api.writer;
+
+import org.apache.flink.core.io.IOReadableWritable;
+import org.apache.flink.runtime.event.AbstractEvent;
+import org.apache.flink.runtime.io.AvailabilityProvider;
+
+import java.io.IOException;
+
+/**
+ * The record writer delegate provides the availability function for task processor, and it might represent
+ * a single {@link RecordWriter} or multiple {@link RecordWriter} instances in specific implementations.
+ */
+public interface RecordWriterDelegate<T extends IOReadableWritable> extends AvailabilityProvider, AutoCloseable {
```

Comment:
We did not explicitly introduce the method for returning this number. During `#getRecordWriter`, different implementation can check whether the index argument is invalid or not. 

ATM I have no obvious preference whether to introduce it and how to use it in practice to get extra benefits.

Reply:
It would only make sense to introduce it if we have any use case where we want to iterate over the delegate without knowing the number of record writers beforehand. 
Afaik we use it only during the initialization of the OperatorChain and during tests. So probably safe to skip now.

Reply:
Exactly, if we need this information in other cases future, we can do that then. 

ATM only the `OperatorChain` relies on other information to get this number. And if we introduce it now, we might involve in the concern that how to guarantee these two informations are strictly consistent. I do not want to touch that trouble now. :(